File: IOTConnectSDK/iotconnect-sdk/src/main/java/com/iotconnectsdk/webservices/CallWebServices.java
package com.iotconnectsdk.webservices;

import com.iotconnectsdk.utils.IotSDKUrls;
import com.iotconnectsdk.utils.IotSDKWsUtils;
import com.iotconnectsdk.webservices.interfaces.WsResponseInterface;
import com.iotconnectsdk.webservices.requestbean.SyncServiceRequest;

import okhttp3.ResponseBody;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class CallWebServices {

    public void sync(String url, final SyncServiceRequest syncServiceRequest, final WsResponseInterface wsResponseInterface) {

        Call<ResponseBody> call = IotSDKWsUtils.getAPIService(url).sync(url, syncServiceRequest);
        call.enqueue(new Callback<ResponseBody>() {
            @Override
            public void onResponse(Call<ResponseBody> call, Response<ResponseBody> response) {
                try {
                    wsResponseInterface.onSuccessResponse(IotSDKUrls.SYNC_SERVICE, response.body().string());
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }

            @Override
            public void onFailure(Call<ResponseBody> call, Throwable t) {
                try {
                    wsResponseInterface.onFailedResponse(IotSDKUrls.SYNC_SERVICE, 0, t.getMessage());
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
        });
    }

    public void getDiscoveryApi(String discoveryUrl, final WsResponseInterface wsResponseInterface) {

        Call<ResponseBody> call = IotSDKWsUtils.getAPIService(discoveryUrl).getDiscoveryApi(discoveryUrl);
        call.enqueue(new Callback<ResponseBody>() {
            @Override
            public void onResponse(Call<ResponseBody> call, Response<ResponseBody> response) {
                try {
                    wsResponseInterface.onSuccessResponse(IotSDKUrls.DISCOVERY_SERVICE, response.body().string());
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }

            @Override
            public void onFailure(Call<ResponseBody> call, Throwable t) {
                try {
                    wsResponseInterface.onFailedResponse(IotSDKUrls.DISCOVERY_SERVICE, 0, t.getMessage());
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
        });
    }

   /* private void showDialog(Activity com.softweb.iotconnectsdk.activity) {
        try {
            progressDialog = new ProgressDialog(com.softweb.iotconnectsdk.activity);
            if (com.softweb.iotconnectsdk.activity != null && !com.softweb.iotconnectsdk.activity.isFinishing()) {
                progressDialog.setMessage(IotSDKConstant.PLEASE_WAIT);
                progressDialog.setCancelable(false);
                progressDialog.show();
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private void hideDialog(Activity com.softweb.iotconnectsdk.activity) {
        if (com.softweb.iotconnectsdk.activity != null && !com.softweb.iotconnectsdk.activity.isFinishing() && progressDialog != null && progressDialog.isShowing()) {
            progressDialog.dismiss();
        }
    }*/

}

File: IOTConnectSDK/iotconnect-sdk_2_1/src/main/java/com/iotconnectsdk/webservices/CallWebServices.kt
package com.iotconnectsdk.webservices

import android.util.Log
import com.iotconnectsdk.utils.IotSDKUrls
import com.iotconnectsdk.utils.IotSDKWsUtils.getAPIService
import com.iotconnectsdk.webservices.interfaces.WsResponseInterface
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import retrofit2.HttpException
import java.net.UnknownHostException

internal class CallWebServices {

    fun sync(url: String?, wsResponseInterface: WsResponseInterface) {

        val service = getAPIService(url)
        CoroutineScope(Dispatchers.IO).launch {
            val response = service?.sync(url)
            try {
                if (response != null) {
                    withContext(Dispatchers.Main) {
                        if (response.isSuccessful) {
                            response.body()?.let {
                                wsResponseInterface.onSuccessResponse(
                                    IotSDKUrls.SYNC_SERVICE,
                                    response.body()!!.string()
                                )
                            }
                        } else {
                            wsResponseInterface.onFailedResponse(
                                IotSDKUrls.SYNC_SERVICE,
                                response.code(),
                                response.message() + " " + response.errorBody()?.string()
                            )
                        }
                    }
                }
            } catch (e: HttpException) {
                Log.e("REQUEST", "Exception ${e.message}")
                wsResponseInterface.onFailedResponse(IotSDKUrls.SYNC_SERVICE, 0, e.message)
            } catch (t: Throwable) {
                wsResponseInterface.onFailedResponse(IotSDKUrls.SYNC_SERVICE, 0, t.message)
            } catch (e: Exception) {
                e.printStackTrace()
                wsResponseInterface.onFailedResponse(IotSDKUrls.SYNC_SERVICE, 0, e.message)
            }
        }

    }

    fun getDiscoveryApi(discoveryUrl: String?, wsResponseInterface: WsResponseInterface) {

        try {
            val service = getAPIService(discoveryUrl)
            CoroutineScope(Dispatchers.IO).launch {
                val response = service?.getDiscoveryApi(discoveryUrl)
                try {
                    if (response != null) {
                        withContext(Dispatchers.Main) {
                            if (response.isSuccessful) {
                                response.body()?.let {
                                    wsResponseInterface.onSuccessResponse(
                                        IotSDKUrls.DISCOVERY_SERVICE,
                                        response.body()!!.string()
                                    )
                                }
                            } else {
                                wsResponseInterface.onFailedResponse(
                                    IotSDKUrls.DISCOVERY_SERVICE,
                                    response.code(),
                                    response.message() + " " + response.errorBody()?.string()
                                )
                            }
                        }
                    } else {

                    }
                } catch (e: HttpException) {
                    Log.e("REQUEST", "Exception ${e.message}")
                    wsResponseInterface.onFailedResponse(IotSDKUrls.DISCOVERY_SERVICE, 0, e.message)
                } catch (t: Throwable) {
                    wsResponseInterface.onFailedResponse(IotSDKUrls.DISCOVERY_SERVICE, 0, t.message)
                } catch (e: Exception) {
                    wsResponseInterface.onFailedResponse(IotSDKUrls.DISCOVERY_SERVICE, 0, e.message)
                    e.printStackTrace()
                }
            }
        } catch (e: UnknownHostException) {
            wsResponseInterface.onFailedResponse(IotSDKUrls.DISCOVERY_SERVICE, 0, e.message)
            e.printStackTrace()
        }
    }
}